master.js

```diff
@@ -832,10 +839,71 @@ class ControlConnection extends BaseConnection {
 		user.roles = resolvedRoles;
 	}
 
+	async setUserAdminRequestHandler(message) {
+		let { name, create, admin } = message.data;
+		let user = db.users.get(name);
+		if (!user) {
+			if (create) {
+				this.user.checkPermission("core.user.create");
+				user = createUser(name);
+			} else {
+				throw new errors.RequestError(`User '${name}' does not exist`);
+			}
+		}
+
+		user.isAdmin = admin;
+		this.broadcastEventToSlaves({ data: { name, admin }}, link.messages.adminlistUpdate);
+	}
+
+	async setUserBannedRequestHandler(message) {
+		let { name, create, banned, reason } = message.data;
+		let user = db.users.get(name);
+		if (!user) {
+			if (create) {
+				this.user.checkPermission("core.user.create");
+				user = createUser(name);
+			} else {
+				throw new errors.RequestError(`User '${name}' does not exist`);
+			}
+		}
+
+		user.isBanned = banned;
+		user.banReason = reason;
+		this.broadcastEventToSlaves({ data: { name, banned, reason }}, link.messages.banlistUpdate);
+	}
+
+	async setUserWhitelistedRequestHandler(message) {
+		let { name, create, whitelisted } = message.data;
+		let user = db.users.get(name);
+		if (!user) {
+			if (create) {
+				this.user.checkPermission("core.user.create");
+				user = createUser(name);
+			} else {
+				throw new errors.RequestError(`User '${name}' does not exist`);
+			}
+		}
+
+		user.isWhitelisted = whitelisted;
+		this.broadcastEventToSlaves({ data: { name, whitelisted }}, link.messages.whitelistUpdate);
+	}
+
 	async deleteUserRequestHandler(message) {
-		if (!db.users.delete(message.data.name)) {
+		let user = db.users.get(message.data.name);
+		if (!user) {
 			throw new errors.RequestError(`User '${message.data.name}' does not exist`);
 		}
+
+		if (user.is_admin) {
+			this.broadcastEventToSlaves({ data: { name, admin: false }}, link.messages.adminlistUpdate);
+		}
+		if (user.is_whitelisted) {
+			this.broadcastEventToSlaves({ data: { name, whitelisted: false }}, link.messages.whitelistUpdate);
+		}
+		if (user.is_banned) {
+			this.broadcastEventToSlaves({ data: { name, banned: false, reason: "" }}, link.messages.banlistUpdate);
```

Comment:
Does this not support ban reason? Seems like something users would expect to have.

Reply:
Ban with ban reason is supported.  What you see here is the notification of removal from the ban list due to the user account being deleted.

Reply:
I see